bindings/csharp/net4.6/Regorus.cs

```diff
@@ -1,27 +1,61 @@
+//-----------------------------------------------------------------------
+// <copyright file="Regorus.cs" company="Microsoft">
+//    Copyright (c)2012 Microsoft. All rights reserved.
+// </copyright>
+// <summary>
+//    Contains code for the Regorus Policy Engine base class.
+// </summary>
+//-----------------------------------------------------------------------
+
+
 using System;
 using System.Text;
+using System.IO;
 
-namespace Regorus
+namespace Microsoft.WindowsAzure.Regorus.IaaS
 {
-    public class Exception : System.Exception
-    {
-	public Exception(string message) : base(message) {}
-    }
-
-    public class Engine : ICloneable
-    {
+	//public class Exception : System.Exception
+	//{
+	//public Exception(string message) : base(message) {}
+	//}
+	public class PolicyEngine : ICloneable, IDisposable
+	{
 	unsafe private RegorusFFI.RegorusEngine* E;
-	public Engine()
+	
+	public PolicyEngine()
 	{
-	    unsafe
+		unsafe
 	    {
 		E = RegorusFFI.API.regorus_engine_new();
 	    }
 	}
 
+	private bool disposed = false;
+
+	public void Dispose()
+	{
+		Dispose(true);
+		GC.SuppressFinalize(this); // Prevent the destructor from being called
+	}
+	
+	protected virtual void Dispose(bool disposing)
```

Comment:
`disposing` is always passed in as true. Therefore, is it needed?

Reply:
One suggestion I have is that you could do something like this (pseudo C#):
```C#
      public void Dispose()
      {
                // Swap null with E in a thread safe manner
		unsafe RegorusFFI.RegorusEngine* ptr = Interlocked.Exchange(ref E, 0);
		if ptr != null {
		    RegorusFFI.API.regorus_engine_drop(E);
                }
		
		GC.SuppressFinalize(this); // Prevent the destructor from being called
	}
```

Reply:
> `disposing` is always passed in as true. Therefore, is it needed?

in destruct function it is passed false.

Reply:
> One suggestion I have is that you could do something like this (pseudo C#):
> 
> ```cs
>       public void Dispose()
>       {
>                 // Swap null with E in a thread safe manner
> 		unsafe RegorusFFI.RegorusEngine* ptr = Interlocked.Exchange(ref E, 0);
> 		if ptr != null {
> 		    RegorusFFI.API.regorus_engine_drop(E);
>                 }
> 		
> 		GC.SuppressFinalize(this); // Prevent the destructor from being called
> 	}
> ```

thanks for the suggestion.

Reply:
@anakrish Exchange function doesn't support pointers very well. Since overall, the code is not thread safe, I think only this function to add thread-safe code is not enough. For the time being, let's not add related code, meanwhile in this case, SuppressFinalize() won't need either. when the thread need arises, adding it altogether. do you agree? Thanks.

Reply:
> Since overall, the code is not thread safe, I think only this function to add thread-safe code is not enough.

Fair point.

>  meanwhile in this case, SuppressFinalize() won't need either.

I'm not that familiar with C# to answer this. 

> when the thread need arises, adding it altogether. do you agree?

Yes, we can add threading related code when need arises. I'd suggest documenting the assumptions about thread-safety somewhere in the code. That will also help the C# binding in upstream Regorus.

Reply:
good. I will add that.

Reply:
I will also restructure the C# binding directory a little bit when preparing my PR to upstream this. this code is customized dotnet4.6. I will create sub directories for dotnet8.0 and above and this one. thanks for the review.